File: apps/robot/robot/sim.py
import rclpy
from geometry_msgs.msg import TransformStamped, Twist
from nav_msgs.msg import Odometry
from rclpy.node import Node
from sensor_msgs.msg import BatteryState, CameraInfo, Image, JointState, NavSatFix, NavSatStatus
from tf2_ros import StaticTransformBroadcaster, TransformBroadcaster

from robot.camera_view import intrinsics, render_field
from robot.kinematics import enu_to_geodetic, integrate_pose, yaw_to_quaternion

MAX_LINEAR_VELOCITY = 0.6
MAX_ANGULAR_VELOCITY = 1.5
WHEEL_RADIUS = 0.178
WHEEL_HALF_TRACK = 0.527
IMAGE_WIDTH = 640
IMAGE_HEIGHT = 400
CAMERA_FOV_DEG = 70.0
CAMERA_HEIGHT = 0.2
LATITUDE_ORIGIN = 45.4215
LONGITUDE_ORIGIN = -75.6972

MAP_FRAME = "map"
ODOM_FRAME = "odom"
BASE_FRAME = "base_link"
CAMERA_FRAME = "camera_optical_frame"


class Sim(Node):
    def __init__(self):
        super().__init__("sim")
        self.x = 0.0
        self.y = 0.0
        self.theta = 0.0
        self.linear_velocity = 0.0
        self.angular_velocity = 0.0
        self.battery_percent = 100.0
        self.left_wheel_angle = 0.0
        self.right_wheel_angle = 0.0
        self.last_update = self.get_clock().now()

        self.odom_pub = self.create_publisher(Odometry, "odom", 10)
        self.image_pub = self.create_publisher(Image, "camera/image_raw", 10)
        self.camera_info_pub = self.create_publisher(CameraInfo, "camera/camera_info", 10)
        self.gps_pub = self.create_publisher(NavSatFix, "gps/fix", 10)
        self.battery_pub = self.create_publisher(BatteryState, "battery", 10)
        self.joint_pub = self.create_publisher(JointState, "joint_states", 10)

        self.tf_broadcaster = TransformBroadcaster(self)
        self.static_tf_broadcaster = StaticTransformBroadcaster(self)
        self.publish_static_transforms()

        self.create_subscription(Twist, "cmd_vel", self.on_cmd_vel, 10)
        self.create_timer(0.02, self.update)
        self.create_timer(1.0 / 15.0, self.publish_camera)
        self.create_timer(0.2, self.publish_gps)
        self.create_timer(1.0, self.publish_battery)
        self.get_logger().info("sim ready, subscribed to /cmd_vel")

    def on_cmd_vel(self, msg):
        self.linear_velocity = max(-MAX_LINEAR_VELOCITY, min(MAX_LINEAR_VELOCITY, msg.linear.x))
        self.angular_velocity = max(-MAX_ANGULAR_VELOCITY, min(MAX_ANGULAR_VELOCITY, msg.angular.z))

    def update(self):
        now = self.get_clock().now()
        dt = (now - self.last_update).nanoseconds / 1e9
        self.last_update = now
        if dt <= 0.0:
            return
        self.x, self.y, self.theta = integrate_pose(
            self.x, self.y, self.theta,
            self.linear_velocity, self.angular_velocity, dt,
        )
        drain = (0.02 + 0.5 * (abs(self.linear_velocity) + abs(self.angular_velocity))) * dt
        self.battery_percent = max(0.0, self.battery_percent - drain)
        self.publish_odometry(now)
        self.publish_base_transform(now)
        left_speed = self.linear_velocity - self.angular_velocity * WHEEL_HALF_TRACK
        right_speed = self.linear_velocity + self.angular_velocity * WHEEL_HALF_TRACK
        self.left_wheel_angle += left_speed / WHEEL_RADIUS * dt
        self.right_wheel_angle += right_speed / WHEEL_RADIUS * dt
        self.publish_joint_states(now)

    def publish_odometry(self, stamp):
        qx, qy, qz, qw = yaw_to_quaternion(self.theta)
        odom = Odometry()
        odom.header.stamp = stamp.to_msg()
        odom.header.frame_id = ODOM_FRAME
        odom.child_frame_id = BASE_FRAME
        odom.pose.pose.position.x = self.x
        odom.pose.pose.position.y = self.y
        odom.pose.pose.orientation.x = qx
        odom.pose.pose.orientation.y = qy
        odom.pose.pose.orientation.z = qz
        odom.pose.pose.orientation.w = qw
        odom.twist.twist.linear.x = self.linear_velocity
        odom.twist.twist.angular.z = self.angular_velocity
        odom.pose.covariance[0] = 0.001
        odom.pose.covariance[7] = 0.001
        odom.pose.covariance[35] = 0.01
        odom.twist.covariance[0] = 0.001
        odom.twist.covariance[35] = 0.01
        self.odom_pub.publish(odom)

    def publish_base_transform(self, stamp):
        qx, qy, qz, qw = yaw_to_quaternion(self.theta)
        transform = TransformStamped()
        transform.header.stamp = stamp.to_msg()
        transform.header.frame_id = ODOM_FRAME
        transform.child_frame_id = BASE_FRAME
        transform.transform.translation.x = self.x
        transform.transform.translation.y = self.y
        transform.transform.rotation.x = qx
        transform.transform.rotation.y = qy
        transform.transform.rotation.z = qz
        transform.transform.rotation.w = qw
        self.tf_broadcaster.sendTransform(transform)

    def publish_joint_states(self, stamp):
        joint_state = JointState()
        joint_state.header.stamp = stamp.to_msg()
        joint_state.name = ["wheel_fl_joint", "wheel_fr_joint", "wheel_rl_joint", "wheel_rr_joint"]
        joint_state.position = [
            self.left_wheel_angle, self.right_wheel_angle,
            self.left_wheel_angle, self.right_wheel_angle,
        ]
        self.joint_pub.publish(joint_state)

    def publish_static_transforms(self):
        map_to_odom = TransformStamped()
        map_to_odom.header.stamp = self.get_clock().now().to_msg()
        map_to_odom.header.frame_id = MAP_FRAME
        map_to_odom.child_frame_id = ODOM_FRAME
        map_to_odom.transform.rotation.w = 1.0
        self.static_tf_broadcaster.sendTransform([map_to_odom])

    def publish_camera(self):
        stamp = self.get_clock().now().to_msg()
        frame = render_field(
            self.x, self.y, self.theta, IMAGE_WIDTH, IMAGE_HEIGHT, CAMERA_FOV_DEG, CAMERA_HEIGHT
        )
        image = Image()
        image.header.stamp = stamp
        image.header.frame_id = CAMERA_FRAME
        image.height = IMAGE_HEIGHT
        image.width = IMAGE_WIDTH
        image.encoding = "rgb8"
        image.is_bigendian = 0
        image.step = IMAGE_WIDTH * 3
        image.data = frame.tobytes()
        self.image_pub.publish(image)
        self.camera_info_pub.publish(self.make_camera_info(stamp))

    def make_camera_info(self, stamp):
        fx, fy, cx, cy = intrinsics(IMAGE_WIDTH, IMAGE_HEIGHT, CAMERA_FOV_DEG)
        info = CameraInfo()
        info.header.stamp = stamp
        info.header.frame_id = CAMERA_FRAME
        info.width = IMAGE_WIDTH
        info.height = IMAGE_HEIGHT
        info.distortion_model = "plumb_bob"
        info.d = [0.0, 0.0, 0.0, 0.0, 0.0]
        info.k = [fx, 0.0, cx, 0.0, fy, cy, 0.0, 0.0, 1.0]
        info.r = [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
        info.p = [fx, 0.0, cx, 0.0, 0.0, fy, cy, 0.0, 0.0, 0.0, 1.0, 0.0]
        return info

    def publish_gps(self):
        latitude, longitude = enu_to_geodetic(self.x, self.y, LATITUDE_ORIGIN, LONGITUDE_ORIGIN)
        fix = NavSatFix()
        fix.header.stamp = self.get_clock().now().to_msg()
        fix.header.frame_id = BASE_FRAME
        fix.status.status = NavSatStatus.STATUS_FIX
        fix.status.service = NavSatStatus.SERVICE_GPS
        fix.latitude = latitude
        fix.longitude = longitude
        fix.altitude = 0.0
        fix.position_covariance = [
            1.0, 0.0, 0.0,
            0.0, 1.0, 0.0,
            0.0, 0.0, 4.0,
        ]
        fix.position_covariance_type = NavSatFix.COVARIANCE_TYPE_DIAGONAL_KNOWN
        self.gps_pub.publish(fix)

    def publish_battery(self):
        battery = BatteryState()
        battery.header.stamp = self.get_clock().now().to_msg()
        battery.voltage = 11.5 + 1.2 * (self.battery_percent / 100.0)
        battery.percentage = self.battery_percent / 100.0
        battery.power_supply_status = BatteryState.POWER_SUPPLY_STATUS_DISCHARGING
        battery.present = True
        self.battery_pub.publish(battery)


def main():
    rclpy.init()
    node = Sim()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: apps/robot/src/sim.rs
use std::{
    sync::Arc,
    time::{Duration, Instant, SystemTime, UNIX_EPOCH},
};

use oxidros::{
    msg::{
        common_interfaces::{
            diagnostic_msgs::msg::{
                DiagnosticArray, DiagnosticStatus, DiagnosticStatusSeq, KeyValue, KeyValueSeq,
            },
            geometry_msgs::msg::Twist,
            nav_msgs::msg::Odometry,
            sensor_msgs::msg::{BatteryState, CameraInfo, Image, NavSatFix, NavSatStatus},
            std_msgs::msg::String as StringMsg,
        },
        msg::{F64Seq, RosString, U8Seq},
    },
    prelude::*,
};
use tokio::time::interval;

use crate::{
    battery_diagnostic_level, command_is_stale, enu_to_geodetic,
    frames::{BASE_LINK, CAMERA_OPTICAL, ODOM},
    integrate_pose, intrinsics, render_field,
};

const IMAGE_WIDTH: usize = 640;
const IMAGE_HEIGHT: usize = 400;
const CAMERA_FOV_DEG: f64 = 70.0;
const CAMERA_HEIGHT: f64 = 0.2;
const LATITUDE_ORIGIN: f64 = 45.4215;
const LONGITUDE_ORIGIN: f64 = -75.6972;
const PHYSICS_HZ: f64 = 50.0;
pub const DEADMAN_SECONDS: f64 = 2.0;
const ROBOT_URDF: &str = include_str!("../urdf/robot.urdf");
const CAMERA_HZ: f64 = 15.0;
const GPS_HZ: f64 = 5.0;
const ODOM_HZ: f64 = 30.0;
const BATTERY_HZ: f64 = 1.0;
const BATTERY_IDLE_DRAIN: f64 = 0.02;
const BATTERY_MOTION_DRAIN: f64 = 0.5;
const BATTERY_MIN_VOLTAGE: f64 = 11.5;
const BATTERY_VOLTAGE_SPAN: f64 = 1.2;

fn now_stamp() -> (i32, u32) {
    let since = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    (since.as_secs() as i32, since.subsec_nanos())
}

#[derive(Default)]
struct SimState {
    x: f64,
    y: f64,
    theta: f64,
    linear: f64,
    angular: f64,
    battery_percent: f64,
}

impl SimState {
    fn new() -> Self {
        SimState {
            battery_percent: 100.0,
            ..Default::default()
        }
    }

    fn set_command(&mut self, twist: &Twist) {
        self.linear = twist.linear.x;
        self.angular = twist.angular.z;
    }

    fn stop(&mut self) {
        self.linear = 0.0;
        self.angular = 0.0;
    }

    fn integrate(&mut self, dt: f64) {
        (self.x, self.y, self.theta) =
            integrate_pose(self.x, self.y, self.theta, self.linear, self.angular, dt);
    }

    fn speed(&self) -> f64 {
        self.linear.abs() + self.angular.abs()
    }

    fn camera(&self) -> (Image, CameraInfo) {
        let (sec, nanosec) = now_stamp();
        let (fx, fy, cx, cy) = intrinsics(IMAGE_WIDTH, IMAGE_HEIGHT, CAMERA_FOV_DEG);
        let frame = render_field(
            self.x,
            self.y,
            self.theta,
            IMAGE_WIDTH,
            IMAGE_HEIGHT,
            CAMERA_FOV_DEG,
            CAMERA_HEIGHT,
        );

        let mut image = Image::new().unwrap();
        image.header.stamp.sec = sec;
        image.header.stamp.nanosec = nanosec;
        image.header.frame_id = RosString::new(CAMERA_OPTICAL).unwrap();
        image.height = IMAGE_HEIGHT as u32;
        image.width = IMAGE_WIDTH as u32;
        image.encoding = RosString::new("rgb8").unwrap();
        image.is_bigendian = 0;
        image.step = (IMAGE_WIDTH * 3) as u32;
        let mut data = U8Seq::new(frame.len()).unwrap();
        data.as_mut_slice().copy_from_slice(&frame);
        image.data = data;

        let mut info = CameraInfo::new().unwrap();
        info.header.stamp.sec = sec;
        info.header.stamp.nanosec = nanosec;
        info.header.frame_id = RosString::new(CAMERA_OPTICAL).unwrap();
        info.width = IMAGE_WIDTH as u32;
        info.height = IMAGE_HEIGHT as u32;
        info.distortion_model = RosString::new("plumb_bob").unwrap();
        info.d = F64Seq::new(5).unwrap();
        info.k = [fx, 0.0, cx, 0.0, fy, cy, 0.0, 0.0, 1.0];
        info.r = [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0];
        info.p = [fx, 0.0, cx, 0.0, 0.0, fy, cy, 0.0, 0.0, 0.0, 1.0, 0.0];

        (image, info)
    }

    fn gps(&self, lat_origin: f64, lon_origin: f64) -> NavSatFix {
        let (sec, nanosec) = now_stamp();
        let (latitude, longitude) = enu_to_geodetic(self.x, self.y, lat_origin, lon_origin);
        let mut fix = NavSatFix::new().unwrap();
        fix.header.stamp.sec = sec;
        fix.header.stamp.nanosec = nanosec;
        fix.header.frame_id = RosString::new(BASE_LINK).unwrap();
        fix.status.status = NavSatStatus::STATUS_FIX;
        fix.status.service = NavSatStatus::SERVICE_GPS;
        fix.latitude = latitude;
        fix.longitude = longitude;
        fix.altitude = 0.0;
        fix.position_covariance = [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 4.0];
        fix.position_covariance_type = NavSatFix::COVARIANCE_TYPE_DIAGONAL_KNOWN;
        fix
    }

    fn odom(&self) -> Odometry {
        let (sec, nanosec) = now_stamp();
        let mut odom = Odometry::new().unwrap();
        odom.header.stamp.sec = sec;
        odom.header.stamp.nanosec = nanosec;
        odom.header.frame_id = RosString::new(ODOM).unwrap();
        odom.child_frame_id = RosString::new(BASE_LINK).unwrap();
        odom.pose.pose.position.x = self.x;
        odom.pose.pose.position.y = self.y;
        odom.pose.pose.orientation.z = (self.theta / 2.0).sin();
        odom.pose.pose.orientation.w = (self.theta / 2.0).cos();
        odom.twist.twist.linear.x = self.linear;
        odom.twist.twist.angular.z = self.angular;
        odom
    }

    fn step_battery(&mut self) -> BatteryState {
        self.battery_percent = (self.battery_percent
            - (BATTERY_IDLE_DRAIN + BATTERY_MOTION_DRAIN * self.speed()))
        .max(0.0);
        let (sec, nanosec) = now_stamp();
        let mut battery = BatteryState::new().unwrap();
        battery.header.stamp.sec = sec;
        battery.header.stamp.nanosec = nanosec;
        battery.voltage =
            (BATTERY_MIN_VOLTAGE + BATTERY_VOLTAGE_SPAN * (self.battery_percent / 100.0)) as f32;
        battery.percentage = (self.battery_percent / 100.0) as f32;
        battery.power_supply_status = BatteryState::POWER_SUPPLY_STATUS_DISCHARGING;
        battery.present = true;
        battery
    }

    fn diagnostics(&self) -> DiagnosticArray {
        let (sec, nanosec) = now_stamp();
        let mut array = DiagnosticArray::new().unwrap();
        array.header.stamp.sec = sec;
        array.header.stamp.nanosec = nanosec;

        let percentage_text = format!("{:.0}%", self.battery_percent);

        let mut kv = KeyValue::new().unwrap();
        kv.key = RosString::new("percentage").unwrap();
        kv.value = RosString::new(&percentage_text).unwrap();
        let values = KeyValueSeq::<0>::from_vec(vec![kv]).unwrap();

        let mut status = DiagnosticStatus::new().unwrap();
        status.level = battery_diagnostic_level(self.battery_percent);
        status.name = RosString::new("battery").unwrap();
        status.hardware_id = RosString::new("simulator").unwrap();
        status.message = RosString::new(&percentage_text).unwrap();
        status.values = values;

        array.status = DiagnosticStatusSeq::<0>::from_vec(vec![status]).unwrap();

        array
    }
}

pub async fn run_simulator(
    node: Arc<Node>,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let mut cmd_vel = node.create_subscriber::<Twist>("cmd_vel", None)?;
    let image_pub =
        node.create_publisher::<Image>("camera/image_raw", Some(Profile::sensor_data()))?;
    let camera_info_pub =
        node.create_publisher::<CameraInfo>("camera/camera_info", Some(Profile::sensor_data()))?;
    let gps_pub = node.create_publisher::<NavSatFix>("gps/fix", Some(Profile::sensor_data()))?;
    let battery_pub = node.create_publisher::<BatteryState>("battery", None)?;
    let diag_pub = node.create_publisher::<DiagnosticArray>("/diagnostics", None)?;
    let odom_pub = node.create_publisher::<Odometry>("odom", None)?;
    let robot_description_pub = node.create_publisher::<StringMsg>("robot_description", None)?;

    let mut param_server = node.create_parameter_server()?;
    {
        let mut params = param_server.params.write();
        params.set_parameter(
            "deadman_seconds".to_string(),
            Value::F64(DEADMAN_SECONDS),
            false,
            Some("Halt the robot if no cmd_vel arrives within this many seconds".to_string()),
        )?;
        params.set_floating_point_range("deadman_seconds", 0.1, 60.0, 0.0)?;
        params.set_parameter(
            "latitude_origin".to_string(),
            Value::F64(LATITUDE_ORIGIN),
            false,
            Some("GPS origin latitude in degrees".to_string()),
        )?;
        params.set_parameter(
            "longitude_origin".to_string(),
            Value::F64(LONGITUDE_ORIGIN),
            false,
            Some("GPS origin longitude in degrees".to_string()),
        )?;
    }
    let params = param_server.params.clone();
    let read_f64 = |name: &str, fallback: f64| -> f64 {
        if let Some(p) = params.read().get_parameter(name) {
            if let Value::F64(v) = p.value {
                return v;
            }
            tracing::warn!("parameter '{name}' is not an f64; keeping {fallback}");
        }
        fallback
    };
    let mut deadman_seconds = read_f64("deadman_seconds", DEADMAN_SECONDS);
    let mut lat_origin = read_f64("latitude_origin", LATITUDE_ORIGIN);
    let mut lon_origin = read_f64("longitude_origin", LONGITUDE_ORIGIN);

    let mut state = SimState::new();
    let mut last_command = Instant::now();
    let dt = 1.0 / PHYSICS_HZ;
    let mut physics_tick = interval(Duration::from_secs_f64(dt));
    let mut camera_tick = interval(Duration::from_secs_f64(1.0 / CAMERA_HZ));
    let mut gps_tick = interval(Duration::from_secs_f64(1.0 / GPS_HZ));
    let mut battery_tick = interval(Duration::from_secs_f64(1.0 / BATTERY_HZ));
    let mut odom_tick = interval(Duration::from_secs_f64(1.0 / ODOM_HZ));

    tracing::info!("driving on cmd_vel -> camera/gps/battery");
    let ctrl_c = tokio::signal::ctrl_c();
    tokio::pin!(ctrl_c);
    loop {
        tokio::select! {
            _ = &mut ctrl_c => break,
            updated = param_server.wait() => {
                match updated {
                    Ok(names) => {
                        for name in &names {
                            tracing::info!("parameter '{name}' updated");
                        }
                        deadman_seconds = read_f64("deadman_seconds", deadman_seconds);
                        lat_origin = read_f64("latitude_origin", lat_origin);
                        lon_origin = read_f64("longitude_origin", lon_origin);
                    }
                    Err(e) => tracing::warn!("parameter service error: {e}"),
                }
            }
            message = cmd_vel.recv() => {
                state.set_command(&message?.sample);
                last_command = Instant::now();
            }
            _ = physics_tick.tick() => {
                if command_is_stale(last_command.elapsed().as_secs_f64(), deadman_seconds) {
                    state.stop();
                }
                state.integrate(dt);
            }
            _ = camera_tick.tick() => {
                let (image, info) = state.camera();
                image_pub.send(&image)?;
                camera_info_pub.send(&info)?;
            }
            _ = gps_tick.tick() => gps_pub.send(&state.gps(lat_origin, lon_origin))?,
            _ = odom_tick.tick() => odom_pub.send(&state.odom())?,
            _ = battery_tick.tick() => {
                battery_pub.send(&state.step_battery())?;
                diag_pub.send(&state.diagnostics())?;
                let mut description = StringMsg::new().unwrap();
                description.data = RosString::new(ROBOT_URDF).unwrap();
                robot_description_pub.send(&description)?;
            }
        }
    }

    tracing::info!("shutting down");
    Ok(())
}
